city/crud.py

```
@@ -0,0 +1,57 @@
+from fastapi import HTTPException
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncSession
+from city import models, schemas
+
+
+async def get_city_by_name(db: AsyncSession, city_name: str):
+    selection = select(models.City).where(models.City.name == city_name)
+    result = await db.execute(selection)
+    return result.scalars().first()
+
+
+async def get_city_by_id(db: AsyncSession, cid: int) -> models.City:
+    selection = select(models.City).where(models.City.id == cid)
+    res = await db.execute(selection)
+    return res.scalars().first()
+
+
+async def get_all_cities(db: AsyncSession) -> list[models.City]:
+    selection = select(models.City)
+    res = await db.execute(selection)
+    return res.scalars().all()
+
+
+async def get_all_cities_dict(db: AsyncSession) -> list[dict[str, str | int]]:
+    cities = await get_all_cities(db)
+    return [{"name": city.name, "id": city.id} for city in cities]
+
+
+async def create_city(db: AsyncSession, city: schemas.CityCreate) -> models.City:
+    created = models.City(
+        name=city.name,
+        additional_info=city.additional_info,
+    )
+    db.add(created)
+    await db.commit()
+    await db.refresh(created)
+    return created
```

When creating a city you rely on the router to prevent duplicates, but the DB enforces a unique constraint on `name`. If an integrity error occurs during `await db.commit()`, it will raise and bubble up as a server error. Consider handling `IntegrityError` (from `sqlalchemy.exc`) and returning a clear HTTPException (400) or let it propagate but document this behavior. This will make the POST /cities behavior more robust and meet the requirement to handle invalid inputs gracefully.